### ISSUE TITLE
Possibility to use `Bugsnag.report` outside a try/rescue block

### ISSUE THREAD

**Issue by rafaqueque** [jarednorman/bugsnag-elixir#19]
Hey,

Is it possible to use `Bugsnag.report` outside a `try/rescue` block? We're building a consumer and we want to log some stuff, but it seems we can only do that with this dirty hack:

```
try do
rescue _ -> Bugsnag.report(%FakeException{foo: "bar"})
end
```

It works that way, but if I just use:

```
Bugsnag.report(%FakeException{foo: "bar"})
```

There's no error and nothing on Bugsnag.

Any hints on why this is happening? The `FakeException` has the structure of a basic exception.

**Comment by jarednorman** [comment]
Sorry, I don't know why this is happening, and am not currently actively developing this package (due to the lack of Elixir in my life these days :disappointed:). I will happily merge a change that fixes this, though. It seems like a valid way to use the package.

**Comment by mschae** [comment]
@rafaqueque Best guess is that you don't have a stack trace.

I'm happy to look into making this more pluggable, give me a few days to carve out time for this.

**Comment by amaranthrose** [comment]
@mschae did you manage to solve this problem? Would be great to use the package the way @rafaqueque described.

**Comment by mschae** [comment]
I have not but to be honest I haven't spent much time researching it. Let me see if I can look into it sometime early this week. If you heaven't heard from me till late this week please ping me, lots on my plate right now. :)

**Comment by ephe-meral** [comment]
@mschae ping - did you have time to look into this already?

**Comment by mschae** [comment]
@ephe-meral @amaranthrose @rafaqueque Way too long since I checked into this. Looking right now.

**Comment by mschae** [comment]
Call me crazy but at least in master this seems to be working fine...

I tried `Bugsnag.report %RuntimeError{message: "testme"}`

It made it to Bugsnag ok...

**Comment by ephe-meral** [comment]
@mschae Crazy! It's true I can reproduce it :)
Maybe something with Elixirs stacktrace was broken and has been fixed in 1.3?

**Comment by mschae** [comment]
Potentially. Could also be that something here changed in the mean time. Glad it works now!

**Comment by ephe-meral** [comment]
@mschae Actually I get the bug again now, but I could trace it: Sometimes, System.stacktrace does not return anything but an empty list. However, when that gets passed to Bugsnag, it sends an HTTP OK, but ignores the messages, since it expects a stacktrace to be there.
Now I don't know how to contiue with that problem... I could temporarily fix it by adding a dummy stacktrace in case there is none given, but that doesnt look like a good solution.
Thoughts?

**Comment by mootpointer** [comment]
The trick would be to grab the trace from where you are and add it to the bugnag call I think. 







On Sat, Aug 13, 2016 at 2:45 AM +1000, "Johanna Appel" <notifications@github.com> wrote:












@mschae Actually I get the bug again now, but I could trace it: Sometimes, System.stacktrace does not return anything but an empty list. However, when that gets passed to Bugsnag, it sends an HTTP OK, but ignores the messages, since it expects a stacktrace to be there.

Now I don't know how to contiue with that problem... I could temporarily fix it by adding a dummy stacktrace in case there is none given, but that doesnt look like a good solution.

Thoughts?

**Comment by ephe-meral** [comment]
@mootpointer So, even if I call `System.stacktrace` in my own code and pass the stacktrace in via Option, it still doesn't produce a Stacktrace with the very first call of that function. (Interestingly, all further calls DO produce a valid stacktrace)

**Comment by mootpointer** [comment]
Hmmm. We should probably not try to generate a trace if we already have one provided. 

Let me look at that. 





On Mon, Aug 15, 2016 at 6:22 PM +1000, "Johanna Appel" <notifications@github.com> wrote:












@mootpointer So, even if I call System.stacktrace in my own code and pass the stacktrace in via Option, it still doesn't produce a Stacktrace with the very first call of that function. (Interestingly, all further calls DO produce a valid stacktrace)

**Comment by ephe-meral** [comment]
It already doesn't. It only generates one itself if you don't pass one in.
See:

https://github.com/jarednorman/bugsnag-elixir/blob/master/lib/bugsnag.ex#L21

**Comment by mootpointer** [comment]
Serves me right for working from memory. 
I'll have a play with this when I get a chance to see what I can see. 






On Mon, Aug 15, 2016 at 6:33 PM +1000, "Johanna Appel" <notifications@github.com> wrote:












It already doesn't. It only generates one itself if you don't pass one in.

See:



https://github.com/jarednorman/bugsnag-elixir/blob/master/lib/bugsnag.ex#L21

**Comment by ephe-meral** [comment]
Okay. So as a temporary fix for myself, I'll fork this and add the dummy stacktrace - until we can figure out how to implement it properly :)

**Comment by ephe-meral** [comment]
On another note, since the explicit call to Bugsnag.report is async (i.e. using spawn) without sending any meaningful result back to the calling process, its really hard to debug whether or not the call was actually successfully submitted... maybe it would be reasonable to make it synchronous by default, so that the caller can control it and wrap it in a Task or something?
Or we could offer separate APIs for synchronous reporting and async reporting

**Comment by mootpointer** [comment]
Yeah. That’s definitely a thing. Maybe pop that into an issue so we can keep track of this (pending a PR)

On Mon, 29 Aug 2016 at 9:15 PM Johanna Appel

<
mailto:Johanna Appel <notifications@github.com>
> wrote:

a, pre, code, a:link, body { word-wrap: break-word !important; }

On another note, since the explicit call to Bugsnag.report is async (i.e. using spawn) without sending any meaningful result back to the calling process, its really hard to debug whether or not the call was actually successfully submitted... maybe it would be reasonable to make it synchronous by default, so that the caller can control it and wrap it in a Task or something?

Or we could offer separate APIs for synchronous reporting and async reporting

—

You are receiving this because you were mentioned.

Reply to this email directly,
https://github.com/jarednorman/bugsnag-elixir/issues/19#issuecomment-243096862
, or

**Comment by grantovich** [comment]
Using `Bugsnag.report` is currently possible outside of `rescue` blocks, and it appears the async reporting issue was fixed by #31, so I'll close this out.
